### ISSUE TITLE
WSL2 support for in Cortex-Debug. Discussion and Strategy

### ISSUE THREAD

**Issue by haneefdm** [Marus/cortex-debug#467]
WSL2 is next on my list. I am trying to set up a Windows machine (my previous WSL2 seems like it got corrupted). I would like people to subscribe to this Issue and comment and help test it. There are several issues and workarounds already related to this and I would like to consolidate the discussion here.

First, comments are welcome on how it should work.

**Comment by haneefdm** [comment]
References: Issues #461, #402, #361, #66 and PR #328. There may be more.

**Comment by s13n** [comment]
Just my $0.02:

I would want to be able to work in either of those two setups:

1. Building takes place _in the WSL2 subsystem directly_. Here, the cross tools (arm-none-eabi-gdb etc.) are installed in the WSL2 subsystem. Since there doesn't seem to be any USB passthrough into WSL2, the debugger, which runs inside the WSL2, needs to establish contact with the probe that is connected to the Windows host using networking.

2. Building takes place _in a docker container_ running within the WSL2 subsystem. You would typically be running Docker Desktop, which on Windows offers support for WSL2. In a nutshell, a Linux container with the cross tools runs within the WSL2 subsystem, and VScode runs on Windows in Dev Container mode. Again, since gdb runs inside the container, the network must be used to contact the debug probe.

The second approach may seem more complex, but it offers the advantage that the tool environment for a project can be shrink-wrapped in a ready-made container that is described concisely with a Dockerfile that can be maintained and versioned in its own git repository. Multiple such containers can be supported in parallel for different projects, without them getting in each other's way.

The problem of how GDB talks to the probe is a multifaceted one, that depends a lot on the actual probe. It will have to be a network connection of some sort. Some example scenarios are:

1. You have a probe with a network port, for example a SEGGER JLink PRO. In this case, there's probably no difference in the communication setup compared to the non-WSL2 case. With SEGGER, you install the Linux version of the SEGGER JLink software in the Dev Container (second case above) or WSL2 (first case above). No special software is needed on the Windows host.

2. You have a SEGGER JLink connected to the Windows host via USB. In this case you can use the JLink remote server included in the SEGGER JLink software installation. Thus, you have to install the Windows version on your host. You only need to run the remote server on the host. Within WSL2 or within the Dev Container, you need the Linux version of the SEGGER software. The JLink GDB server is run in WSL2 or the Dev Container, respectively, and it is configured to connect to the remote server via IP. See the SEGGER docs for that. This scenario can also be used with a SEGGER probe connected to a different computer on the network. If your 3rd party probe can be reflashed to act as a JLink probe, which includes numerous probes integrated on an evaluation board, this should work, too.

3. You run the GDB server on the Windows host, and have GDB inside WSL2 or the Dev Container contact this GDB server using IP. You might have to run the GDB server manually, unless a way can be found to start it from within the Dev Container. This should work even without a JLink based probe. In WSL2 or the Dev Container, no special probe-related software should be needed.

The first step would be to offer information on how to configure those scenarios properly, i.e. to describe what should work or what shouldn't. In this step it would be acceptable having to start some server manually, instead of having everything happening automatically when the debug button is pressed. It would also be acceptable to enter IP addresses or other settings manually depending on the local setup.

The second step would be to automate this stuff as much as possible.

Does this help you, @haneefdm ?

**Comment by haneefdm** [comment]
> Does this help you, @haneefdm ?

@s13n  Oh, helps a LOT, and thank you for such detail. And, a lot to think about over the weekend.

Would you agree with the following?

1. Hard constraint: gdb should run wherever the source is compiled and ideally, this would be in WSL2/Linux environment. Or else path-names in the ELF file will be messed up. Messy, but can be corrected with [gdb source-paths](https://sourceware.org/gdb/current/onlinedocs/gdb/Source-Path.html).
2. Hard constraint:  the gdb-server should run wherever the HW/probe is. This, to accommodate all (most) gdb-servers. J-Link PRO's remote-server is an exception but in truth, the real probe SW is running where the probe is right?

**Comment by s13n** [comment]
I would be able to live with constraint 1, but perhaps others will disagree.

I think constraint 2 is a bit too restrictive. I would say GDB server should run either where the probe is, or where GDB is. Supporting a third location would be overkill. This also supports the JLink remote server. One instance where it matters is when you use a JLink with network interface, for example the JLink Pro. You want to run the JLink GDB server where the GDB is, otherwise you end up installing the SEGGER software on both the host and in the dev container.

Of course, with a USB connected probe, you have to install the respective software on the host.

**Comment by haneefdm** [comment]
I added Constraint 2 because of the physical interface (USB) the HW connects to. This is more of a constraint to **me** as I **have** to find a solution that works in that way. And, I am not saying supporting a 3rd location at all -- sorry, if my words implied that. To summarize, the gdb-server and the HW under debug are always on the same machine, attached at the hip, so to say.

It is also a generalization, beyond WIndows+WSL2.

**Comment by s13n** [comment]
Well, USB isn't the only physical interface that is supported by probes, it is only the most common one (by far). Why would you restrict yourself like that? What do you gain?

**Comment by haneefdm** [comment]
I am not restricting myself to it. It is a reality that I am stating. Once the server is not local, then it is remote. There is no in-between and I don't see WSL as an in-between thing. Maybe this is where I am wrong. I really don't care how the gdb-server connects to the HW.

We got the following scenarios

1. gdb-server runs where VSCode/gdb runs. Great, we already do that
2. gdb-server runs on another machine. Okay, not great. TCP port selection, launching of the server is not automatic and until recently, SWO was not supported. This works 100% but, *people are not happy with this.*
3. gdb-server cannot run locally in some cases which is where the WSL situation comes in but it degenerates to scenario 2 above.

It is 2/3 that we are trying to address to make it look almost like # 1. Note that we NEVER talk to the gdb-server which is why we don't care where it lives. 90+% of the time TCP ports are used and sometimes serial ports. But someone has to launch it. We never had to worry if the gdb-server is using USB or some other communication mechanism.

The reason GDB is run where the compiling is done is because of pathnames embedded in the elf file. If these are not right, then breakpoints don't work, stack traces will not have references back to source code, etc.

One thing we did not talk about is where is VSCode running? In my head, gdb and VSCode are running on the same machine. One reason is that all communication with GDB happens over stdio, while gdb itself may talk to the gdb-server over some connection (local or remote). Another is that this is the model GDB has chosen and has worked for over 3 decades.

Btw, I the MB failed on the PC where I used to have WSL installed. It was my only PC. I ordered a PC and waiting. I normally do my testing in a VM but here that gets convoluted. Mac running Win in VM which in turn hosting WSL2. So, no experiments until next week.

**Comment by s13n** [comment]
Have a look at https://code.visualstudio.com/docs/remote/remote-overview for some overview of how VS Code is used in remote mode, which includes WSL2.

We are talking about a scenario where VS Code runs on Windows in remote mode. The remote OS is either the WSL2 subsystem directly, or a docker container running within WSL2. The VS Code Server runs there, the source code resides there, and the cross tools including GDB run there.

The WSL2 case has the feature that you can launch Windows software from within the WSL2 subsystem. AFAIK you can't do this from within a docker container, but maybe the fact that VS Code runs on Windows provides a way to start some process there, even though the actual debugging takes place in the remote OS under control of the VS Code Server. I have no idea if VS Code helps you with that.

I am relatively new with the dev container way of working with VS Code, but I already prefer working in this way, due to the simplicity of maintaining a build environment that is specific to a project. You can also run the same container on a remote machine, for example your build server, rather than locally within WSL2, and you shouldn't notice much of a difference.

**Comment by haneefdm** [comment]
> Have a look at https://code.visualstudio.com/docs/remote/remote-overview for some overview of how VS Code is used in remote mode, which includes WSL2.

Okay, that is a very different model. I was aware of that. I will look into it while I wait, but if you see the repo for it, it has 802 issues, no commits since 5/11/2021.  It is a package containing 3 extensions. Last commit was 4 months ago in those repos.

Some of it is the inverse model of what I was thinking.

One thing that is important to me (selfishly) is how to debug this extension itself. Without that it would be horrible.

I worked a bit on the MS C++ debug adapter and it was very difficult to do any cross-platform debugging of the extension. It was like a one-man circus show to juggle multiple VSCodes running on different machines. Both VSCode and Visual Studio were needed. I can't even explain.

**Comment by s13n** [comment]
When using Docker Desktop for Windows, you can have the host's IP address resolved from within the container by using `host.docker.internal` as the DNS name.

See https://docs.docker.com/desktop/windows/networking/

**Comment by haneefdm** [comment]
Oh, that is super nice with Docker. Thanks @s13n

**Comment by wbober** [comment]
@haneefdm I think looking into `/etc/resolv.conf` is the correct way. It does work for me:

![image](https://user-images.githubusercontent.com/221354/131322963-e991fae0-873c-434f-b4aa-3b10c7700c2e.png)

**Comment by haneefdm** [comment]
@wobe, Thanks. Does that work if the host Windows is using VPN. Don't you see many entries in the `/etc/resolve.conf`. The solution I am thinking may not need to know the host-ip at all. VSCode may help in this regard.

**Comment by wbober** [comment]
> @wobe, Thanks. Does that work if the host Windows is using VPN. Don't you see many entries in the `/etc/resolve.conf`. The solution I am thinking may not need to know the host-ip at all. VSCode may help in this regard.

Don't see much of a difference, i.e., I have the same contents in the `resolv.conf` file.

**Comment by GitMensch** [comment]
>     1. Hard constraint: gdb should run wherever the source is compiled and ideally, this would be in WSL2/Linux environment. Or else path-names in the ELF file will be messed up. Messy, but can be corrected with [gdb source-paths](https://sourceware.org/gdb/current/onlinedocs/gdb/Source-Path.html).

I "vote" against that constraint. Just allow setting source-path directly in the settings and everything is fine.

**Comment by haneefdm** [comment]
Technically it is not a hard constraint for me. Of course, you can use source paths just like you can today. It has to do with the client-server VSCode architecture. This extension and the GDB are attached at the hip. That is the true hard constraint for me.

![image](https://user-images.githubusercontent.com/41269583/133800490-3d822b80-0cc9-4a7e-8cfd-d4decbc88bae.png)

https://code.visualstudio.com/docs/remote/remote-overview.

As things stand, in one incarnation, Cortex-Debug would be classified in the "Remote OS" box (WSL, Docker, etc.) while the GUI itself would be in the "Local OS" box. That picture is not totally applicable to what we are doing, btw. Especially for where the 'Application is running'. You can also see where the Source code box is.

That little green box on the bottom right of the picture above is GDB.

I don't even know if that arch. is feasible for me but it is a start as a lot of groundwork has already been laid out.

**Comment by GitMensch** [comment]
That only applies if your use VS Code Server - and as those binaries are non-free I don't use that. That's likely the reason why I commonly think of "Local OS with source and GDB", attached to "GDB Server with the process". This "simple" scenario also works quite fine since years for most setups.
Note: when running on Windows MSYS2 provides a gdb-multiarch.exe - so the "Debugger" part is solved (objdump not).

**Comment by haneefdm** [comment]
> and as those binaries are non-free I don't use that.

Which binaries? The VSCode server(s)? And free as in Open Source or some other meaning (costs)?

In VSCode's mind 'Local OS' means the host running the UI. To me, 'Local OS' also meant "Local OS with source and GDB" but I had to teach myself a different way of thinking. VSCode is my host so, playing by the host's rules and thus terminology. I have to back and edit all my comments to make sure.

**Comment by GitMensch** [comment]
Yes, the vscode servers. It is all about freedom, not price. This actually leads to vscode server not running everywhere I'd like it to for remote debugging (I'm not sure if it actually works in every distro one can install in wsl).
... actually the client part in vscode is also closed source and the "gratis" extensions needed to work for that: are only licensed to be used on "Visual Studio Code binaries" (= the one provided by MS where you are even not allowed to distribute own copies.... - and even from a practical view: those binaries are not available for all GNU/Linux distros where vscode actually runs - which is the reason that I only use binaries that are as free as vscode [the main source], nowadays mainly VSCodium).

**Comment by GitMensch** [comment]
Note: in VSCode's mind "Local OS" is also something that run's vscode in a browser: it is actually the UI.

**Comment by lagerholm** [comment]
Maybe this: 
https://www.xda-developers.com/wsl-connect-usb-devices-windows-11/
https://www.elevenforum.com/t/connecting-usb-devices-to-wsl.2514/
will change some of the scope. 
Not tested yet so I haven't verified that it actually works to have the probes connected directly to WSL2 with USB.

**Comment by haneefdm** [comment]
@lagerholm Thank you so much for the info. This is great.

**Comment by DanieleNardi** [comment]
Hello,
I tried to use Cortex-Debug in WSL2 using the usbipd-win tool to connect host-connected J-Link to WSL2 environment. J-Link connection looks fine (at first look, it seems it's required Jlink drivers to be installed on both host and wsl2 sides and aligned to the same version), but extension doesn't stop at main.
launcher.json looks like the following:
```
"name": "DVC TopRow Emerald Inventory",
"type": "cortex-debug",
"request": "launch",
"cwd": "${workspaceFolder}",
"armToolchainPath": "/opt/gcc-arm-none-eabi/bin",
"executable": "path/to/elf",
"serverpath": "/opt/SEGGER/JLink/JLinkGDBServerCLExe",
"servertype": "jlink",
"device": "MK10DX256xxx7",
"interface": "jtag",
"serialNumber": "proper serial number",
"runToMain": true,
"stopAtEntry": true,
"svdFile": "path/to/svd"
```

**Comment by TorBlackbeard** [comment]
>  (at first look, it seems it's required Jlink drivers to be installed on both host and wsl2 sides and aligned to the same version

That makes me suspicious. Are you sure you don't get one of the windows-side tools called by accident?
If the USB is properly visible in the WSL, it should work 100% on the linux toolchain?
(all imho, of cause)

Would like to test it out soon!

**Comment by DanieleNardi** [comment]
> That makes me suspicious. Are you sure you don't get one of the windows-side tools called by accident?

I thought the same, I checked, but it wasn't.
> If the USB is properly visible in the WSL, it should work 100% on the linux toolchain? (all imho, of cause)

Same, but actually didn't. Can't say why.

**Comment by GitMensch** [comment]
> Same, but actually didn't. Can't say why.

Are you sure that this is WSL2? If not please check and if necessary convert, then check again.

**Comment by DanieleNardi** [comment]
I checked with usbipd-win team and the communication is fine.

**Comment by JojoS62** [comment]
are you using Windows 11 already? There WSL has more new features that are not available in Win 10.

**Comment by DanieleNardi** [comment]
> are you using Windows 11 already? There WSL has more new features that are not available in Win 10.

No, still Win10. It's a company PC, I don't know when (if) we'll get the update.

**Comment by wbober** [comment]
I have the `usbipd` a go. I'm on Windows 10 and did a simple test: flash a sample program on nRF52840 DK that outputs logs through a JLINK VCOM. Here is the result:

Putty on Windows:
![image](https://user-images.githubusercontent.com/221354/145232197-cfc58fa1-ed27-4002-9f8b-642fe3ee0569.png)

Picocom on WSl2 through `usbipd`:
![image](https://user-images.githubusercontent.com/221354/145232257-ad14dd87-47c3-4789-b807-08857e6d2f5f.png)

As you see there are some errors in the log output on WSL2. I suspect this might related to the issue @DanieleNardi has. If there are communication errors Jlink might not work properly.

**Comment by RoboGnome** [comment]
It's great to see that WSL2 support is beeing pushed.

I'm interested in the following setup:

- The host is a Windows machine
- Everything is installed inside WSL2 (toolchain, openocd, JLink)
- VS Code runs remotetly in WSL2 and debugging is done with tools installed under WSL2

I managed to get a working debugging session using WSL2, USB Passthrough with USB-IP and JLink installed in WSL2.
It worked under Windows 10 (21H2) and Windows 11.
I had some problems with the udev service not running on WSL2, maybe that's also your problem.
Issuing the following commands before attaching the USB device to WSL2 solved my problems:
```
sudo service udev restart
sudo udevadm control --reload
```

You also need to have the jlink udev rules copied to /udev/rules.d.

I documented the process for my EFR32 board here:
[https://github.com/RoboGnome/efr32_base](https://github.com/RoboGnome/efr32_base)

**Comment by GitMensch** [comment]
> I managed to get a working debugging session using WSL2, USB Passthrough with USB-IP and JLink installed in WSL2.

Out of interest: Did you use _any_ proprietary Microsoft extensions or "servers" (installed in WSL or in vscode) or was this _just_ your setup noted above (which distro btw?) plus this extension?

**Comment by wbober** [comment]
@RoboGnome I have the same versions but on Windows 10. The UART doesn't work reliably for me, unfortunately. I'm pretty sure that it's not udev issue.

**Comment by RoboGnome** [comment]
@wbober Well maybe it's dependend on the device or usb port used? But I'm no expert on USB.

The experience for me is stellar and it's exactly what I was missing in WSL2 to do all my embedded development on Windows+WSL2. So thanks to anyone who made it possible.

But I only tested it my Thunderboard Sense 2 and only with a CLI from a example. So I didn't printed a lot of data to see if the console can't keep up. But while testing it, I didn't notice any lag.
The board comes with a JLink CDC UART and after I attach it to my WSL2 I could connect to it using `screen /dev/ttyACM0 115200`. And it worked while I was debugging. Which is neat, because now I finally have a integrated serial console while debugging.  
But I hope that the timing issues get fixed in the future.

**Comment by mpekurin** [comment]
I'm not sure if this makes sense but as a temporary solution the extension can use a gdb server that's running in Windows while everything else is running in WSL. I've tested this setup a bit and it works pretty well. While using usbipd-win looks better in general, it requires running powershell in administrator mode to attach a device, and I'm not sure an extension can do this.

So in theory this should work:
1. Add a new property in `launch.json` for a user to configure where to run the gdb server. Let's call it `runServerInWindows` for now. Write in the tip that it works only for WSL.
2. Detect if the extension is running in WSL via `vscode.env.remoteName`.
3. If the extension is running in WSL and `runServerInWindows` is `true` then the extension runs `gdbserverexecutable.exe` which launches the server on Windows side (any `*.exe` called in WSL is run in Windows). In case `runServerInWindows` is `false`, the extension runs `gdbserverexecutable` which launches the server on WSL side (e.g. if a user uses usbipd-win).
4. Get the ip address to connect to the gdb server. Microsoft writes that [this](https://docs.microsoft.com/en-us/windows/wsl/networking#accessing-windows-networking-apps-from-linux-host-ip) should work but it didn't for me. I had to use the local ip of my machine (192.168.0.xxx) for this.
5. Connect and debug.

**Comment by askariz** [comment]
> WSL2 is next on my list. I am trying to set up a Windows machine (my previous WSL2 seems like it got corrupted). I would like people to subscribe to this Issue and comment and help test it. There are several issues and workarounds already related to this and I would like to consolidate the discussion here.
> 
> First, comments are welcome on how it should work.


Window allow to execute openocd.exe on WSL2 which to access STLINK hardware device on Windows 
Here is my command to access stm device on WSL2 and everything go right 

`
/mnt/c/openocd/bin/openocd.exe -c "gdb_port 50000" -c "tcl_port 50001" -c "telnet_port 50002" -s /mnt/c/Users/Administrator/Desktop/work/ODrive/Firmware  -f interface/stlink.cfg -f target/stm32f7x.cfg -c "reset_config none separate"
Open On-Chip Debugger 0.11.0+dev-00572-g94e7535be (2022-02-19-14:36)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : DEPRECATED target event trace-config; use TPIU events {pre,post}-{enable,disable}
none separate

Info : Listening on port 50001 for tcl connections
Info : Listening on port 50002 for telnet connections
Info : clock speed 2000 kHz
Info : STLINK V2J29M18 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.224258
Info : stm32f7x.cpu: Cortex-M7 r1p1 processor detected      
Info : stm32f7x.cpu: target has 8 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f7x.cpu on 50000
Info : Listening on port 50000 for gdb connections 
`

But when I launch with VScode with launch config 
`
 {
            // For the Cortex-Debug extension
            "type": "cortex-debug",
            "servertype": "openocd",
            "request": "attach",
            "name": "Debug ODrive v4.x - ST-Link",
            "executable": "${workspaceRoot}/build/ODriveFirmware.elf",
            "searchDir": ["/home/askariz/.vscode-server/extensions/marus25.cortex-debug-1.2.2/support/"],
            "configFiles": [
                "interface/stlink.cfg",
                "target/stm32f7x.cfg",
                "openocd-helpers.tcl"
            ],
            
            "showDevDebugOutput": "both",
            "openOCDLaunchCommands": [
                "reset_config none separate"
            ],
            "svdFile": "${workspaceRoot}/Private/v4/STM32F722.svd",
            "cwd": "${workspaceRoot}"
        },
`
It automatically add 
`
/mnt/c/openocd/bin/openocd.exe -c "gdb_port 50000" -c "tcl_port 50001" -c "telnet_port 50002" -s /home/askariz/.vscode-server/extensions/marus25.cortex-debug-1.2.2/support/ -f /home/askariz/.vscode-server/extensions/marus25.cortex-debug-1.2.2/support/openocd-helpers.tcl -f interface/stlink.cfg -f target/stm32f7x.cfg -f openocd-helpers.tcl -c "reset_config none separate"
Open On-Chip Debugger 0.11.0+dev-00572-g94e7535be (2022-02-19-14:36)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
embedded:startup.tcl:26: Error: Can't find /home/askariz/.vscode-server/extensions/marus25.cortex-debug-1.2.2/support/openocd-helpers.tcl
in procedure 'script' 
at file "embedded:startup.tcl", line 26
[2022-02-20T03:41:55.658Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session closed
GDB server session ended. This terminal will be reused, waiting for next session to start...
`

which cause I can't find  Can't find /home/askariz/.vscode-server/extensions/marus25.cortex-debug-1.2.2/support/openocd-helpers.tcl  **But This File Truely exsist on my WSL2...**

Any suggestion ?

**Comment by mpekurin** [comment]
@askariz you need to use `"servertype": "external"` to acces the server that's running on Windows side from WSL2:
```json
"servertype": "external",
"gdbTarget": "hostname.mshome.net:port"
```
where `hostname` is the name of your PC.
For this to work you need to add the firewall rule to allow connection from WSL2 if you haven't:
```powershell
New-NetFirewallRule -DisplayName "WSL" -Direction Inbound -InterfaceAlias "vEthernet (WSL)" -Action Allow
```

**Comment by GitMensch** [comment]
This may be useful for both the WSL scenario and for `external` debugging where the source path on the client (vscode) is not the same as the debugger (gdbserver, possibly running on wsl) sees:

I suggest to add a [`sourceFileMap`](https://code.visualstudio.com/docs/cpp/launch-json-reference#_sourcefilemap) setting. This would need to applied anywhere the client passes paths (breakpoints, "go to cursor") and when the client receives paths (I _guess_ that's the stack only).

Should that be moved out to a separate issue?

**Comment by haneefdm** [comment]
People are using their own gdb-init scripts to add source maps. No help from extension needed. The bigger problem is how breakpoints are set and how stack traces are interpreted. Disassembly is affected as well.. How we look for static variables is also affected -- as in we have to implement what gdb does. VSCode manages breakpoints with paths (file/line) that it sees. It wasn't clear to me how gdb handled file/line breakpoints and if clients (us) have to do the reverse mapping. Yes, a separate issue would be good but also with details of what it all means. In fact, it is more useful in a non WSL setting.

cpptools/cppdbg added [support for source maps](https://code.visualstudio.com/docs/cpp/launch-json-reference#_sourcefilemap). We can check to see what all they do with that info.

**Comment by aaronf-at-d5** [comment]
I've recently been trying to solve the same for a client project.  I read this entire thread and did some experimentation and here's what I came up with.

In my `.bashrc` file in WSL, I added the following line:
```
export HOST_IP=`ip route|awk '/^default/{print $3}'`
```
In the `launch.json` file for VS Code, I added the following to `serverArgs`:
```
"-select", "ip=${env:HOST_IP}"
```

Then, I just need to make sure the JLink Remote Server is started in Windows and running in LAN mode before launching a debug session.  This could probably also be triggered from `launch.json`.  I need to look more into that.  The JLink Remote Server does not let me select the network interface when running in LAN mode and for some reason always picks a VMWare network interface (192.168.253.1).  Somehow this works, even though the JLink GDB Server running in Ubuntu is connecting to 172.28.208.1, which is nice because I don't have to manually copy any IP addresses.

Note, that `HOST_IP` here refers to the Windows host (what VS Code calls the "Local OS" when running a remote connection).

**Comment by aaronf-at-d5** [comment]
As far as hard constraint 1 above
> gdb should run wherever the source is compiled and ideally, this would be in WSL2/Linux environment

I haven't seen any discussion yet about [`-fdebug-prefix-map`](https://gcc.gnu.org/onlinedocs/gcc/Debugging-Options.html#Debugging-Options).  I have had good success with this moving ELF files between separate build and debug environments.  This may not be a good all around solution though as it is gcc-specific.

**Comment by sullivanmj** [comment]
It may be possible to attach USB-based debuggers to WSL via cortex-debug, that does not require any of the following:

1. Use of USB-IP
2. Pre-launch/post-launch tasks to start remote GDB server
3. Running of other remote debugging assistant applications such as J-Link Remote Server

My thought is to leverage the fact that [Windows executables, when invoked from WSL, execute within the context of Windows](https://learn.microsoft.com/en-us/windows/wsl/filesystems#run-windows-tools-from-linux).

This means that we can invoke a GDB server on the Windows host using the cortex-debug `serverPath` and perhaps `serverArgs` (to allow remote connections, etc) debug attributes.

Then the GDB client could be invoked from within WSL and attached per usual, but with some of the tricks listed above to use the correct target IP address and port - either using something like `host.docker.internal:50000` or `hostname.mshome.net:50000`, or something like `${env:HOST_IP}:50000`.

Debugging would commence as usual, and any other clients of the GDB server that you wish to attach from Windows could also operate normally. Then at the end of the debug session, the GDB connection would be cleanly closed as usual from the GDB client.

To achieve all of this, I was hoping to find a way to use both the `serverPath` and `gdbTarget` debug attributes, but it seems that `gdbTarget` is only used when the `servertype` is `external`, in which case, the `serverPath` is not used. This is an understandable behavior given the context of normal GDB client/server connections, but it prevents the approach I'm describing from being used.

I'm thinking that the simplest way to work around that limitation, would be the creation of an `overrideInitCommands` debug attribute. This attribute would allow manual specification of how the connection from the GDB client to the server should be initialized from within launch.json. I have a branch that attempts this functionality located [here](https://github.com/sullivanmj/cortex-debug/tree/override-init-commands). Presumably, this approach could be leveraged in a variety of other use-cases, too. It would be used in the `launch.json` like so:

`"overrideInitCommands":"target remote ${env:HOST_IP}:50000"`

Am I off my rocker? Is there a better way of achieving what I've described above with the existing functionality of cortex-debug, or should I open a PR?

**Comment by aaronf-at-d5** [comment]
@sullivanmj, I like where you're going with this line of thinking.  The reason using `preLaunchTask` to forward to `tasks.json` works is because a lot of the Windows-native paths are automatically mapped into WSL with the path renaming necessary to point into `/mnt/c/`.  This allows us to just invoke `JLinkGDBServerCL.exe` and lot Windows hooks take care of it.  The process is clunky though and fails more easily.  It would be nice to tell the Cortex Debug extension directly in `launch.json`, that we're using the Windows executable.  You would still have to provide the `-nolocalhostonly` in `serverArgs` and use the `${env:HOST_IP}:2331` trick.

**Comment by sullivanmj** [comment]
@aaronf-at-d5 Good point about the path being mapped into WSL. I agree, it would be great to have this way of doing it natively supported by Cortex Debug. I now that I wrote that post, it appears that I didn't actually discover this first, @askariz appears to have done this with openocd. But it looks like they were invoking that separately from the launch.json.

**Comment by mpekurin** [comment]
@sullivanmj 

> To achieve all of this, I was hoping to find a way to use both the `serverPath` and `gdbTarget` debug attributes, but it seems that `gdbTarget` is only used when the `servertype` is `external`, in which case, the `serverPath` is not used. This is an understandable behavior given the context of normal GDB client/server connections, but it prevents the approach I'm describing from being used.

That's what `preLaunchTask` is used for, to bypass that behavior and basically do what the `serverPath` should do. So maybe a simpler alternative to your solution would be just adding a support of using both `serverPath` and `gdbTarget`. Then a config like this will do the trick:

```json
{
  "type": "cortex-debug",
  "name": "Launch",
  "request": "launch",
  "servertype": "theServerOfYourChoice",
  "serverPath": "/mnt/c/path/to/theServerOfYourChoice.exe",
  "serverArgs": ["--port", "61234"],
  "gdbTarget": "${env:HOSTNAME}.mshome.net:61234"
}
```

Additionally, if this is (or will be) required, the extension can check if `serverPath` is an `.exe` to understand where the server is run.

> Is there a better way of achieving what I've described above with the existing functionality of cortex-debug, or should I open a PR?

Without using USB-IP or tasks, most likely no.

**Comment by romancardenas** [comment]
Hi, folks!

I'm setting WSL2 with Cortex-Debug for a course and I was wondering what is the approach you recommend at this point. Should I...?
- Use USB-IP and expose my board in WSL
- or trying to do as @mpekurin mentioned and somehow running OpenOCD on windows but connecting to a GDB server running on WSL2?

I'm not expert in VSCode plugins nor WSL, but I can try to open a PR with this change

**Comment by GitMensch** [comment]
I'm no export in anything of those but would suggest to connect to WSL2 first (with either the SSH or WSL2 extensions; either from https://github.com/jeanp413 or, if you prefer non-free "in preview" extensions with telemetry the matching one from Microsoft) and fully debug there, with the possible need to expose your board to WSL first.

**Comment by haneefdm** [comment]
I am getting reports that WSL2 with USB-IP works very well. In this case, you would install OpenOCD/Jlink/whatever right on the Linux side and use the Linux versions.

But, make sure that your USB is working first by using various command-line tools (lsusb). You may run into firewall issues but there are workarounds for those too. Let Google be your friend. Make sure you can launch the gdb-server (OpenOCD, JLink, etc.) from the command-line first and that it recognizes your adapter/probe.

**Comment by romancardenas** [comment]
It does work pretty well! The only issue is that students need to handle the USB attachment etc., which might be a bit confusing for them.

So far I'll use the usbipd approach. If I have time I'll try to do a PR with the required changes to skip all this process and leave openOCD run on Windows while GDB is on WSL.

**Comment by haneefdm** [comment]
@romancardenas About making the PR, you also have to handle allocating/finding the TCP ports if you are running on the Windows side. I think running the server on the Linux side is still a better option. We want things like SWO, RTT, etc. to also work. Some instructions for your students may be beneficial.

**Comment by szszoke** [comment]
I managed to work around the problem of J-Link software not being able to resolve the IP address `host.docker.internal`. I needed this to be able to connect to a J-Link remote server that was running on the same host as my Dev Container. Hard-coding an IP like `172.17.0.1` did not work because different host platforms have different IP addresses.

In order for this to work, you need to run docker with the following argument: `--add-host=host.docker.internal:host-gateway`. This argument is technically redundant on Windows and Mac, but it is needed on Linux. You can do this by adding the following to`devcontainer.json`:

```json
{
  "runArgs": ["--add-host=host.docker.internal:host-gateway"]
}
```

This will ensure that you have `host.docker.internal` entry in `/etc/hosts` which points to the IP of your host that is running the Dev Container.

The second step is to resolve this hostname to an IP so that we can pass it to `JLinkGDBServer`.

This can be done with the following command:

```bash
$ getent hosts host.docker.internal
172.17.0.1      host.docker.internal
```

We can further clean this command up with the help of `awk`:


```bash
$ getent hosts host.docker.internal | awk '{ print $1 }'
172.17.0.1
```

Great, now we have the IP address and it is not hard-coded. Unfortunately there is obvious way to provide this IP to the debug configuration without hard-coding it so we need a different solution.

I noticed that there is a `serverpath` parameter which allows us to provide a custom path for `JLinkGDBServer`. This is what allowed me to finally make things working. I created a shell script with the following content:

```bash
#!/bin/sh

REMOTE_JLINK_IP=`getent hosts host.docker.internal | awk '{ print $1 }'`

JLinkGDBServer -select ip=$REMOTE_JLINK_IP "$@"
```

What this will do is to get the IP address behind the `host.docker.internal` entry in `/etc/hosts` and store it in `REMOTE_JLINK_IP`. I then just invoke `JLinkGDBServer`, pass the IP via `-select ip=$REMOTE_JLINK_IP` and pass any other arguments that were passed to the script via `"$@"`.

I saved this script in my `.devcontainer` directory as `remote-gdb-server.sh` and then updated my debug configuration to use it by adding `"serverpath": "./.devcontainer/remote-gdb-server.sh"`.

Don't forget to run `chmod +x remote-gdb-server.sh`.

This exact pattern can also be used in a `Makefile`:

```makefile
REMOTE_JLINK_IP := $(shell getent hosts host.docker.internal | awk '{ print $$1 }')
```

One word of caution however is that `$` needs to be escaped, which becomes `$$`.

**Comment by haneefdm** [comment]
I wish someone volunteers to document what works in our Wiki. We could have sections for WSL, docker, etc. I can start a Wiki Page if you want but I am not good for the content. You all know way more than me.

**Comment by szszoke** [comment]
> I can start a Wiki Page if you want but I am not good for the content.

If you start a Wiki page, I could contribute my previous comment as an article about using J-Link from a Docker container.

**Comment by haneefdm** [comment]
https://github.com/Marus/cortex-debug/wiki/Working-with-WSL2-or-Docker

Feel free to re-organize the page. I just started something. Maybe we want two separate pages, we can decide later.

**Comment by szszoke** [comment]
I think my comment would fit in better with the things in `J Link Specific Configuration`, since it is highly specific to that programmer and the software that comes with it.

**Comment by haneefdm** [comment]
Make sections and sub-sections as you please. I would expect some setup differences for various gdb-servers.

**Comment by szszoke** [comment]
I think I will be using my setup for a bit to iron out some kinks. Until then, if somebody needs help, hopefully they see my previous comment.

**Comment by rhempel** [comment]
I have had some success with the following scenario:

- Locked down corporate Windows laptop - but with wsl and docker allowed
- VSCode on Windows side
- Generic project with .devcontainer and Dockerfile to set up an Ubuntu 22 container suitable for development
- Project specific Docker volume for the actual project - works on Windows and Linux hosts and is fast, rebuilding machine doesn't kill work in progress
- STLink detected with USB-IP and passed to WSL first, then start Docker instance in --privileged mode
- VSCode as editor and debugger works

It works, but single stepping is painful, and I'm not sure I like the requirement to mess with USB-IP so I will check some of the comments above and rather start st-util on the host machine and connect to the gdb-server running on the host from the container where gdb is running.

Is anyone interested in my progress on this?

**Comment by jnz86** [comment]
> Is anyone interested in my progress on this?

I'm following. I'd be more interested had I not given up and just switched to linux desktop for at LEAST until this is all solved for Windows, which may be never.

**Comment by rhempel** [comment]
I have a potential solution here:

https://github.com/rhempel/docker-devcontainers/tree/main/adaptabuild-example

The difference from my original scenario is that now we run stlink-util on the Windows host, and get the correct IP address into the container `/etc/hosts` file using this line in the `devcontainer,json` file:

```
// This line is the key to accessing the GDB server on the host machine
// when running Docker on a Windows machine.
//
// See .vscode/launch.json for details

  "runArgs": [
    "--add-host=host.gdb.gateway:host-gateway",
  ],
```

In `.vscode/launch.json` the key is:

```
      "gdbTarget": "host.gdb.gateway:4242",
```

Note, the container is based on Ubuntu 22.04 and pulls in everything needed to build a simple application, plus Shpinx for docs and gcov. The `launch.json` file assumes a project called adaptabuild-example which you can build in ~/projects.

Note: you might need to change ownership on ~/projects to `adaptabuild:adaptabuild` to allow cloning there.

The adaptabuild-example project is here:

https://github.com/rhempel/adaptabuild-example

I'm still busy getting it polished and writing a tutorial, but please let me know if this works for you :-)

**Comment by CaptainKAZ** [comment]
https://devblogs.microsoft.com/commandline/windows-subsystem-for-linux-september-2023-update/
It seems that we can use new "mirror" networking mode for connecting gdb server running on windows and gdb client on wsl2?
I tried but openocd claims that it fails to bind tcl port.

```
openocd.exe -c "gdb_port 50000" -c "tcl_port 50001" -c "telnet_port 50002" -s /home/neo/Robocon2024-Embedded-BoardC -f /home/neo/.vscode-server/extensions/marus25.cortex-debug-1.12.1/support/openocd-helpers.tcl -f openocd.cfg -c "CDRTOSConfigure auto" -c CDLiveWatchSetup
xPack Open On-Chip Debugger 0.12.0+dev-01312-g18281b0c4-dirty (2023-09-04-22:32)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
CDLiveWatchSetup
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
/home/neo/.vscode-server/extensions/marus25.cortex-debug-1.12.1/support/openocd-helpers.tcl: stm32f4x.cpu configure -rtos auto
/home/neo/.vscode-server/extensions/marus25.cortex-debug-1.12.1/support/openocd-helpers.tcl: Info: Setting gdb-max-connections for target 'stm32f4x.cpu' to 2
Error: couldn't bind tcl to socket on port 50001: No error
```

**Comment by xpz24** [comment]
I believe wsl, cortex debug and openocd works well together now, at least for nucleo32 boards over usb. I tried it on a Windows 11, WSL2, ubuntu 23.10, 6.2.x kernel, usbipd nightly version. Before that I updated the ST-Link firmware to the latest version.
https://steelph0enix.github.io/posts/vscode-cubemx-setup/ this guide helped a lot.

I used the openocd, customised version by STMicroelectronics, the one in ubuntu repos would work?
STMCubeMX (linux version, with latest JRE, most likely unnecessary, the bundled JRE would be ok I think) to generate the code.
stm32-for-vscode by dingen automated most of the stuff in the guide.
I used the svd file from STM website.
Programing and flashing works  well, Debugging also works well I say, can see the variables, peripherals, registers and memory. 

All possible from wsl side, only windows program I had to install was usbipd.

**Comment by AaronFontaine-DojoFive** [comment]
@xpz24, I think that much of the discussion here has been concerned with how to bridge out of WSL for the debugger. A lot of that has focused on getting the gdb server launched on the Windows side and then getting a WSL-side gdb client connected to it. I now think the better approach is, as you say, to use usbipd. There is now even GUI for managing this. Goliath has a good [article](https://blog.golioth.io/usb-support-in-wsl2-now-with-a-gui/) about it.

**Comment by rhempel** [comment]
I have tried both scenarios, and unless I am missing something, running the gdb-server over usbipd is al least an order of magnitude slower than gdb-server running on the host and connecting over a socket.

Plus, running gdb-server on the host and using a socket works on Windows/Linux and MacOS if there is a gdb-server for your debugger hardware on MacOS

**Comment by BenSchoelerSonos** [comment]
> I have a potential solution here:
> 
> https://github.com/rhempel/docker-devcontainers/tree/main/adaptabuild-example
> 
> The difference from my original scenario is that now we run stlink-util on the Windows host, and get the correct IP address into the container `/etc/hosts` file using this line in the `devcontainer,json` file:
> 
> ```
> // This line is the key to accessing the GDB server on the host machine
> // when running Docker on a Windows machine.
> //
> // See .vscode/launch.json for details
> 
>   "runArgs": [
>     "--add-host=host.gdb.gateway:host-gateway",
>   ],
> ```
> 
> In `.vscode/launch.json` the key is:
> 
> ```
>       "gdbTarget": "host.gdb.gateway:4242",
> ```
> 
> Note, the container is based on Ubuntu 22.04 and pulls in everything needed to build a simple application, plus Shpinx for docs and gcov. The `launch.json` file assumes a project called adaptabuild-example which you can build in ~/projects.
> 
> Note: you might need to change ownership on ~/projects to `adaptabuild:adaptabuild` to allow cloning there.
> 
> The adaptabuild-example project is here:
> 
> https://github.com/rhempel/adaptabuild-example
> 
> I'm still busy getting it polished and writing a tutorial, but please let me know if this works for you :-)

Hi @rhempel, did you happen to write a tutorial for this method?  Very interested :)

**Comment by rhempel** [comment]
Well, sort of ...

The main repository for getting started is here: https://github.com/rhempel/docker-devcontainers

That contains a folder for building a Docker base image for embedded development (it's about 2.2 GB and takes 5 minutes or so with typical dev PC and 50 mbps internet. It works on Windows and Linux and so far one M1 Mac :-)

https://github.com/rhempel/docker-devcontainers/tree/main/docker-base-images

The README.md file covers some background and has instructions for kicking of the build of the Docker image. I highly recommend we stick with the 22.04 variant due to some issues building rpi2040 with the gcc 13 compiler - something about TLS initialization.

Next, you want to build the actual devcontainer - thats in the adaptabuild-example-22.04 folder here:

https://github.com/rhempel/docker-devcontainers/tree/main/adaptabuild-example-22.04

Make sure the Docker daemon is running before going to the next step 

Just rightclick and open with VSCode, and VSCode will see the .devcontainer folder and try to build and start a shiny new container. Note that it will also create a Docker Volume called  adaptabuild-volume and it will mount (not copy) your local user's .ssh file to the container root - that makes git work transparently :-) It will also automagically pull in all the VSCode extensions needed for embedded ARM work.

If you get that far - congratulations - you will be presented with two shell terminals in VSCode. Now what?

Navigate to ~/projects and:

git clone https://github.com/rhempel/adaptabuild-example
cd adaptabuild-example
git submodule update --init

That should pull in the latest example of my adaptabuild example code and framework.

Let me know if I messed up the URLs (again) I think I have everything set to https:

Clcik on Terminal/RunTask and choose "Build All" - that should build all the targets curently supported

You're almost there! On your host machine (WIndows/Mac/Linux) you will run st-util in a terminal - that will look for your ST-Link compatible dev board and set up a GDB server waiting on port 4242

Now click on Launch in VSCode - you should see options for a couple of ST and nrf and rpi boards - clicking the launch for the board you have should start the debugger - connect to the STlink over a network socket - and bingo - you are done.

For anyone interested in testing this, I would be more than happy to support your questions and can arrange a Zoom call to help. I am interested in feedback for this setup - it sounds complex but it can get a new developer up and running in under 30 minutes, and I am very intetrested in collecting user feedback so that I can improve it.

I won't give private contact info here but you can find me on LinkedIn - unfortunately there are two Ralph Hempel users, I'm the one that matches my avatar here and exLEGO :-)

**Comment by rhempel** [comment]
Sorry for the long post above - here is a ( hopefully ) better summary:

https://github.com/rhempel/adaptabuild-example

I would be happy to get feedback - let me know in the comments and we can set up a 30 minute screen share.

**Comment by JanekGraff** [comment]
Hey, sorry re-activate this thread, but i just wanted to give my 2 cents to this since i stumbled upon this while setting up a debug environment in VSCode WSL for an STM32U5. 

I managed to get this running using basically the default launch.json configuration which the STM32 VSCode extension provides. I had to adjust the `executable` Path since i'm using a non-standard build system and the `armToolchainPath` and `gdbPath`. I used [usbipd](https://github.com/dorssel/usbipd-win) to pass through the ST-Link device to WSL (see [here](https://github.com/dorssel/usbipd-win/wiki/WSL-support) for more detailed instructions).

Here is my launch config if anyone is interested:

```json
        {
            "name": "Build & Debug Microcontroller - ST-Link",
            "cwd": "${workspaceFolder}",
            "type": "cortex-debug",
            "executable": "${workspaceFolder}/out/my-executable.elf",
            "request": "launch",
            "servertype": "stlink",
            "device": "STM32U5A5ZJTx", //MCU used
            "interface": "swd",
            "serialNumber": "",        //Set ST-Link ID if you use multiple at the same time
            "runToEntryPoint": "main",
            "svdFile": "${config:STM32VSCodeExtension.cubeCLT.path}/STMicroelectronics_CMSIS_SVD/STM32U5A5.svd",
            "v1": false,               //Change it depending on ST Link version
            "serverpath": "${config:STM32VSCodeExtension.cubeCLT.path}/STLink-gdb-server/bin/ST-LINK_gdbserver",
            "stm32cubeprogrammer":"${config:STM32VSCodeExtension.cubeCLT.path}/STM32CubeProgrammer/bin",
            "stlinkPath": "${config:STM32VSCodeExtension.cubeCLT.path}/STLink-gdb-server/bin/ST-LINK_gdbserver",   
            "armToolchainPath": "${env:HOME}/bin/arm-gnu-toolchain-14.2.rel1-x86_64-arm-none-eabi/bin",
            "gdbPath":"${env:HOME}/bin/arm-gnu-toolchain-14.2.rel1-x86_64-arm-none-eabi/bin/arm-none-eabi-gdb",            
            "serverArgs": [
                "-m","0",
            ],
            "preLaunchTask": "Build + Flash"
        },
````

Here is the tasks config containing the `Build + Flash` task (note the `"reset=HWrst"` argument is not default by the STM32 extension, but seems to work more reliable in my case. I also had to adjust the CubeProgrammer path in the `"command"` field.):

```json
    "tasks": [
        {
            "type": "shell",
            "label": "CubeProg: Flash project (SWD)",
            "command": "${config:STM32VSCodeExtension.cubeCLT.path}/STM32CubeProgrammer/bin/STM32_Programmer_CLI",
            "args": [
                "--connect",
                "port=swd",
                "reset=HWrst",
                "--download",
                "${workspaceFolder}/out/my-executable.elf",
                "-hardRst", // Hardware reset - if rst pin is connected
                "-rst", // Software reset (backup)
                "--start" // Start execution
            ],
            "options": {
                "cwd": "${workspaceFolder}"
            },
            "problemMatcher": []
        },
        {
            "label": "Build + Flash",
            "dependsOrder": "sequence",
            "dependsOn": [
                "Make: clean rebuild",
                "CubeProg: Flash project (SWD)",
            ]
        },
                {
            "type": "shell",
            "label": "Make: clean rebuild",
            "command": "make clean && make debug -j 16",
            "group": "build",
            "problemMatcher": [],
            "detail": "Make template clean rebuild task"
        },
```

Again sorry for re-activating this thread and also sorry if something like this has already been posted, i haven't read through the entire thread. Just thought this might be useful for some folk like me also stumbling over this thread.

**Comment by haneefdm** [comment]
Your docker config needs to install ALL required extensions. VSCode does not do this for you unlike when using a desktop. Not sure why. When docker spins up, all the following should also be installed.

https://github.com/Marus/cortex-debug/blob/9fb01a6c729d9e5fca3eff650281359d8216b9db/package.json#L3104

This requirement is mentioned somewhere in the VSCode docs (not specific to this extension)

**Comment by stemschmidt** [comment]
Mmmh, sorry, not really I guess... I am running it from WSL2 and the thread was pretty close to what I was thought might give me a solution. But thanks for spending time on it! I really like the plugin, it is a great extension to work with.

**Comment by obirchman** [comment]
Started JLinkGDBServer in windows and then set these three configs in launch.json.
```
"servertype": "external",
"serverpath": "/opt/SEGGER/JLink/JLinkGDBServerCLExe",
"gdbTarget": "localhost:2331",
```
 It connects, flashes and runs a bit but then there is a fault in the timer IRQ for the HAL tick. In Ozone it runs without issues, might just be missing some initialization / configuration I need from Ozone. I'll see if I can figure it out.

I saw mostly discussion surrounding STLink so thought I'd share. Does anyone know if using external servertype instead of JLink could cause any issues? Will RTT work anyway for example?

I'll update if I make some progress.

**### Full launch config**
```
    {
      "name": "Main - J-Link - WSL",
      "type": "cortex-debug",
      "request": "launch",
      "servertype": "external",
      "device": "STM32H725VG",
      "interface": "swd",
      "cwd": "${workspaceFolder}",
      "executable": "${workspaceFolder}/build/main/main.elf",
      "svdFile": "${workspaceFolder}/utility/STM32H725.svd",
      "serverpath": "/opt/SEGGER/JLink/JLinkGDBServerCLExe",
      "gdbPath": "${workspaceFolder}/../tools/stm32cubeclt_1.18.0/GNU-tools-for-STM32/bin/arm-none-eabi-gdb",
      "gdbTarget": "localhost:2331",
      "rtos": "FreeRTOS",
      "runToEntryPoint": "main",
      "showDevDebugOutput": "raw",
      "rttConfig": {
        "enabled": true,
        "address": "auto",
        "searchId": "SEGGER RTT",
        "decoders": [
          { "port": 0, "type": "console", "label": "RTT" }
        ]
      }
    },
```

Being able to do debugging with the JLink from WSL without forwarding usb ports is enticing!
